docs/garbage-collection.md

```diff
@@ -0,0 +1,133 @@
+
+
+## Garbage Collection in the Registry
+
+Filesystem layers are stored by their content address in the Registry.  This
+has many advantages, one of which is that data is stored once and referred to by manifests.
+See [here](docs/compatibility.md#content-addressable-storage-cas) for more details.
+
+Layers are therefore shared amongst manifests; each manifest maintains a reference
+to the layer.  As long as a layer is referenced by one manifest, it cannot be garbage
+collected.
+
+Manifests and layers can be 'deleted` with the registry API (refer to the API
+documentation [here](docs/spec/api.md#deleting-a-layer) and
+[here](docs/spec/api.md#deleting-an-image) for details).  This API removes references
+to the target and makes them eligible for garbage collection.  It also makes them
+unable to be read via the API.
+
+If a layer is deleted it will be removed from the filesystem when garbage collection
+is run.  If a manifest is deleted the layers to which it refers will be removed from
+the filesystem if no other manifests refers to them.
+
+
+### Example
+
+In this example manifest A references two layers: `a` and `b`.  Manifest `B` references
+layers `b` and `c`.  In this state, nothing is eligible for garbage collection:
+
+```
+A -----> a <----- B
+    \--> b     |
+         c <--/
+```
+
+Manifest B is deleted via the API:
+
+```
+A -----> a     B
+    \--> b
+         c
+```
+
+In this state layer `c` no longer has a reference and is eligible for garbage
+collection.  Layer `a` had one reference removed but will not be garbage
+collected as it is still referenced by manifest `A`.  The blob representing
+manifest `B` will also be eligible for garbage collection.
+
+After garbage collection has been run manifest `A` and its blobs remain.
+
+```
+A -----> a
+    \--> b
+```
+
+
+## How Garbage Collection works
+
+Garbage collection runs in two phases. First, in the 'mark' phase, the process
+scans all the manifests in the registry. From these manifests, it constructs a
+set of content address digests. This set is the 'mark set' and denotes the set
+of blobs to *not* delete. Secondly, in the 'sweep' phase, the process scans all
+the blobs and if a blob's content address digest is not in the mark set, the
+process will delete it.
+
+
+> **NOTE** You should ensure that the registry is in read-only mode or not running at
+> all. If you were to upload an image while garbage collection is running, there is the
+> risk that the image's layers will be mistakenly deleted, leading to a corrupted image.
+
+This type of garbage collection is known as stop-the-world garbage collection.  In future
+registry versions the intention is that garbage collection will be an automated background
+action and this manual process will no longer apply.
+
+
+
+# Running garbage collection
+
+You can run garbage collection by running
+
+`docker run --rm <registry-image-name> garbage-collect /path/to/config.yml`
```

Comment:
Can we provide instructions for running this outside a container as well. For an example command seems lacking some needed details such as mounting volumes, stopping registry container or putting it in read-only mode.